### DISCUSSION TITLE
Stack too deep error with forge coverage , even with --via-ir , optimizer and --ir-minimum flag

### DISCUSSION THREAD

**Discussion by shisukeUrahara** [foundry-rs/foundry#8766] [Help] [answered]
I have a big foundry project and it compiles well and the test cases also run fine.

(i) Running forge coverage gives me this error 

Compiler run failed:
Error: Compiler error (C:\Users\circleci\project\libyul\backends\evm\AsmCodeGen.cpp:67):Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local 
variables. When compiling inline assembly: Variable value0 is 1 slot(s) too deep inside the stack. Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
CompilerError: Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables. When compiling inline assembly: Variable value0 is 1 slot(s) too deep inside the stack. Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.

(ii) running forge coverage --ir-minimum gives me this error 

Warning! "--ir-minimum" flag enables viaIR with minimum optimization, which can result in inaccurate source mappings.
Only use this flag as a workaround if you are experiencing "stack too deep" errors.
Note that "viaIR" is only available in Solidity 0.8.13 and above.
See more: https://github.com/foundry-rs/foundry/issues/3357
[⠊] Compiling...
[⠘] Compiling 165 files with Solc 0.8.23
[⠰] Solc 0.8.23 finished in 41.07s
Compiler run successful with warnings:
Warning (3420): Source file does not specify required compiler version! Consider adding "pragma solidity ^0.8.23;"
--> contracts/paymaster/VerifyingPaymaster.sol

Warning (3420): Source file does not specify required compiler version! Consider adding "pragma solidity ^0.8.23;"
--> test/VerifyingSingletonPaymaster.test.sol

Warning (3420): Source file does not specify required compiler version! Consider adding "pragma solidity ^0.8.23;"
--> test/tokenpaymaster/paymastertest.t.sol

Warning (5667): Unused function parameter. Remove or comment out the variable name to silence this warning.
   --> contracts/base/ExecutionHelper.sol:111:33:
    |
111 |     ) internal virtual returns (bool success, bytes memory result) {
    |                                 ^^^^^^^^^^^^

Analysing contracts...
The application panicked (crashed).
Message:  byte index 1707 is not a char boundary; it is inside '─' (bytes 1706..1709) of `// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.23;

import {PackedUserOperation} from "../../interface/IAccount.sol";

library UserOperationLib {
    /*
        In calldata, the data structure of the UserOperation is always:
           `[...]
Location: crates\evm\coverage\src\analysis.rs:478

This is a bug. Consider reporting it at https://github.com/foundry-rs/foundry

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   1: hid_write<unknown>
      at <unknown source file>:<unknown line>
   2: hid_write<unknown>
      at <unknown source file>:<unknown line>
   3: hid_write<unknown>
      at <unknown source file>:<unknown line>
   4: BaseThreadInitThunk<unknown>
      at <unknown source file>:<unknown line>
   5: RtlUserThreadStart<unknown>
      at <unknown source file>:<unknown line>

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.

How do i fix it ?

I even have the optimizers enabled in the foundry.toml file 

[profile.default]
src = 'contracts'
out = 'out'
libs = ['node_modules', 'lib']
test = 'test'
cache_path  = 'cache_forge'

solc = "0.8.23"
optimizer = true
optimizer_runs = 1000000
via_ir = true
evm_version = "paris"

**Comment by urosognjenovic** [comment]
Did you declare a lot of local variables before the ```value0``` variable?

**Comment by zerosnacks** [comment, marked as answer]
Forge coverage currently is required to disable optimizations in order to accurately track coverage as any optimizations make it so that the AST does not accurately map to the source code anymore. We are adding a warning to `forge coverage` to make users aware of this limitation: https://github.com/foundry-rs/foundry/issues/8840

In the future we may consider an entire rewrite of `forge coverage` that would eliminate the need but this is obviously a huge task only viable post-1.0
